lib/classes/organisations_user_builder.rb

```diff
@@ -11,91 +13,149 @@ class OrganisationsUserBuilder
+
+      organisations_user.position                   = @position unless @position.blank?
+      organisations_user.status                     = @status unless @status.blank?
+      organisations_user.preferred_contact_number   = @preferred_contact_number unless @preferred_contact_number.blank?
+
+      organisations_user.save! if organisations_user.changed?
+      organisations_user
+    end
   end
 
-  def build_user(obj)
-    @user = obj.first_or_initialize(@user_attributes)
-    assign_user_app_acessor
-    @user.save
-    @user
+  # ------------------------
+  # Write methods
+  # ------------------------
+
+  def apply_user_attributes!(user, user_params)
+    [:first_name, :last_name, :email, :mobile, :title].each do |field|
+      user[field] = user_params[field] if user_params[field].present?
+    end
+
+    user.save! if user.changed?
   end
 
-  def assign_user_app_acessor
-    @user.request_from_stock = (@app_name == STOCK_APP)
-    @user.request_from_browse = (@app_name == BROWSE_APP)
+  # ------------------------
+  # Helpers
+  # ------------------------
+
+  def create_params
+    {
+      user_id:                  @user_id,
+      organisation_id:          @organisation_id,
+      position:                 @position,
+      status:                   @status,
+      preferred_contact_number: @preferred_contact_number
+    }
   end
 
-  def update
-    assign_user_app_acessor
-    return fail_with_error(update_user["errors"]) if update_user && update_user["errors"]
+  def notify_user(user)
+    TwilioService.new(user).send_welcome_msg
+  end
 
-    @organisations_user.update(position: @position, preferred_contact_number: @preferred_contact_number)
-    return_success.merge!("organisations_user" => @organisations_user.reload)
+  def super_user?(user)
+    user&.api_user? || user&.has_permission?("can_manage_organisations_users")
   end
 
-  private
+  # ------------------------
+  # Error Management
+  # ------------------------
 
-  def organisation
-    @organisation ||= Organisation.find_by_id(@organisation_id)
+  def strict_find!(model, id)
+    record = id ? model.find_by(id: id) : nil
+    raise Goodcity::BadOrMissingRecord.new(model) if record.blank?
+    record
   end
 
-  def update_user
-    @user.roles << charity_role unless @user.roles.include?(charity_role)
-    if @user.update(@user_attributes)
-      @user
-    else
-      fail_with_error(@user.errors.full_messages.join(" "))
-    end
+  def assert_integrity!(organisations_user)
+    raise Goodcity::ReadOnlyFieldError.new(:user_id).with_status(403)          if organisations_user.user_id != @user_id
+    raise Goodcity::ReadOnlyFieldError.new(:organisation_id).with_status(403)  if organisations_user.organisation_id != @organisation_id
   end
 
-  def user_belongs_to_organisation(user)
-    user.organisation_ids.include?(@organisation_id)
-  end
+  def assert_permissions!
+    return if super_user?(@change_author)
```

Comment:
`super_user` might feel as a strong word 🤔 
can have something simple as valid_user?(@change_author) since we are mainly checking for `can_manage_organisations_users`

Reply:
I think `valid_user?` is equally confusing. But I think `manager?` would work well here : )